File: src/test/java/com/lalkalol/mujahid/commands/impl/ConnectionCommandsTest.java
package com.lalkalol.mujahid.commands.impl;

import com.lalkalol.mujahid.commands.CommandTestFixtures;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class ConnectionCommandsTest {

    @Test
    void joinCommand_requiresMemberVoiceChannel() {
        var h = CommandTestFixtures.harness().memberNotInVoice();

        new JoinCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("voice channel"));
    }

    @Test
    void joinCommand_connectsAndRegistersManager() {
        var h = CommandTestFixtures.harness().memberInVoice(5L, "Lounge");

        new JoinCommand().execute(h.context());

        verify(h.audioController).connect(org.mockito.ArgumentMatchers.any());
        verify(h.lavalink).getOrCreate(CommandTestFixtures.GUILD_ID);
        assertTrue(h.lastReplyDescription().contains("Lounge"));
    }

    @Test
    void leaveCommand_rejectsWhenBotNotConnected() {
        var h = CommandTestFixtures.harness().botNotInVoice();

        new LeaveCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("not connected"));
    }

    @Test
    void leaveCommand_stopsPlaybackAndDisconnects() {
        var h = CommandTestFixtures.harness().botInVoice(1L, "General");
        when(h.lavalink.getExisting(CommandTestFixtures.GUILD_ID)).thenReturn(h.musicManager);

        new LeaveCommand().execute(h.context());

        verify(h.scheduler).stop();
        verify(h.audioController).disconnect(h.guild);
        verify(h.lavalink).destroy(CommandTestFixtures.GUILD_ID);
        assertTrue(h.lastReplyDescription().contains("Bye"));
    }
}

File: src/test/kotlin/com/lalkalol/mujahid/commands/impl/ConnectionCommandsTest.kt
package com.lalkalol.mujahid.commands.impl

import com.lalkalol.mujahid.commands.CommandTestFixtures
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class ConnectionCommandsTest {
    @Test
    fun joinCommand_requiresMemberVoiceChannel() = runBlocking {
        val h = CommandTestFixtures.harness().memberNotInVoice()

        JoinCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("voice channel"))
    }

    @Test
    fun joinCommand_connectsAndRegistersManager() = runBlocking {
        val h = CommandTestFixtures.harness().memberInVoice(5L, "Lounge")

        JoinCommand().execute(h.context())

        verify(h.audioController).connect(any())
        verify(h.lavalink).getOrCreate(CommandTestFixtures.GUILD_ID)
        assertTrue(h.lastReplyDescription().contains("Lounge"))
    }

    @Test
    fun leaveCommand_rejectsWhenBotNotConnected() = runBlocking {
        val h = CommandTestFixtures.harness().botNotInVoice()

        LeaveCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("not connected"))
    }

    @Test
    fun leaveCommand_stopsPlaybackAndDisconnects() = runBlocking {
        val h = CommandTestFixtures.harness().botInVoice(1L, "General")
        whenever(h.lavalink.getExisting(CommandTestFixtures.GUILD_ID)).thenReturn(h.musicManager)

        LeaveCommand().execute(h.context())

        verify(h.scheduler).stop()
        verify(h.audioController).disconnect(h.guild)
        verify(h.lavalink).destroy(CommandTestFixtures.GUILD_ID)
        assertTrue(h.lastReplyDescription().contains("Bye"))
    }
}
